Documentation/UPI/libvirt/libvirt.md

```diff
@@ -0,0 +1,148 @@
+![Direct kernel boot](Images/direct_kernel_boot.png)
+
+**NOTE:** CoreOS installer automatically reboot the server after installation, so when the installation of Fedora CoreOS is finished shut down the VMs, otherwise the installer will keep installing the OS.
+
+After the installation disable the Direct Kernel Boot option, as we don't need it anymore.
+
+At this point you can add the VMs MAC address to the network configuration in libvirt, just like the example [above](#configure-libvirt-network). Stop and start the network after that:
+```
+virsh# net-destroy openshift
+virsh# net-start openshift
+```
+
+### Start the bootstrap server
+After every server in your cluster was provisioned, start the bootstrap server.
+By default, Fedora CoreOS will install the OS from the official OSTree image, so we have to wait a few minutes for the machine-config-daemon to pull and install the pivot image from quay.io. This image is necessary for the kubelet service, as the official Fedora CoreOS image does not include hyperkube.
+After the image was pulled and installed, the server will be rebooted by itself.
+When the server is up again wait for the API service and the MachineConfig service to be spawned (check for the ports 6443 and 22623). Check also for the status of the `bootkube.service`.
+
+### Start the other servers
+**NOTE:** You can start every server in the cluster in the same time of the boostrap server, as they will still waiting for the latter to expose the Kubernetes and MachineConfig API ports. These steps were separated just for convenience.
+
+Now that the bootstrap server is ready, you can start every server of your cluster.
+Just like the bootstrap server, the control planes and the workers will boot with the official Fedora CoreOS image, that does not contains hyperkube. Since hyperkube is missing the kubelet service will not start and so the cluster bootstrapping.
+Wait for the machine-config-daemon to pull the same image as the bootstrap server.
+The servers will reboot themselves and after that they will try to join the cluster, starting the bootstrapping process.
+After reboot `rpm-ostree status` should show something like:
+```
+[core@okd4m0 ~]$ rpm-ostree status
+State: idle
+AutomaticUpdates: disabled
+Deployments:
+● pivot://quay.io/openshift/okd-content@sha256:830ede6ea29c7b3e227261e4b2f098cbe2c98a344b2b54f3f05bb011e56b0209
+              CustomOrigin: Managed by machine-config-operator
+                 Timestamp: 2019-11-15T18:25:12Z
+
+  ostree://fedora:fedora/x86_64/coreos/testing
+                   Version: 31.20191217.2.0 (2019-12-18T14:11:27Z)
+                    Commit: fd3a3a1549de2bb9031f0767d10d2302c178dec09203a1db146e0ad28f38c498
+              GPGSignature: Valid signature by 7D22D5867F2A4236474BF7B850CB390B3C3359C4
+```
+
+
+Wait for the finish of the installation and then login to the console with the kubeadmin credentials, or configure an additional identity provider.
```

Comment:
* use `openshift-installer wait-for bootstrap-complete`
* approve node CSR via `oc get csr -ojson | jq -r '.items[] | select(.status == {} ) | .metadata.name' | xargs --no-run-if-empty oc adm certificate approve`
* `oc patch configs.imageregistry.operator.openshift.io cluster --type merge --patch '{"spec":{"managementState":"Managed","storage":{"emptyDir":{}}}}'` to set internal registry storage to emptyDir (not sure if this is required anymore)
* `openshift-installer wait-for install-complete` to wait for all operators to complete

Reply:
For the third point, I think it would be better if in the long term we add a reference to a storage-focused guide that explains how to configure your storage backend, if you want something more specific than `emptyDir`.
Talking about me, I never configured the image-registry operator for not having a persistent storage.
Even in my non-production installations I always configure a persistent storage, by configuring a NFS server on my host OS.

I always considered `emptyDir` as something to use when you want to create a temporary cluster. But this is just my point of view.

Reply:
Right, lets add a note about this:

----
By default registry would expect a storage provider to provide an RWX volume, see https://docs.openshift.com/container-platform/4.2/registry/configuring-registry-storage/configuring-registry-storage-baremetal.html. Use the following command to use `emptyDir` instead:
`oc patch configs.imageregistry.operator.openshift.io cluster --type merge --patch '{"spec":{"managementState":"Managed","storage":{"emptyDir":{}}}}'`

Reply:
Should be fixed with 0b9ec15

Reply:
Fixed some indentation errors with 03ae740